pydicom/fileset.py

```diff
@@ -0,0 +1,2742 @@
+        yield from self._instances
+
+    def __len__(self) -> int:
+        """Return the number of instances in the File-set."""
+        return len(self._instances)
+
+    def load(
+        self,
+        ds_or_path: Union[Dataset, str, PathLike],
+        include_orphans: bool = True,
+        raise_orphans: bool = False,
+    ) -> None:
+        """Load an existing File-set.
+
+        Existing File-sets that do not use the same directory structure as
+        *pydicom* will be staged to be moved to a new structure. This is
+        because the DICOM Standard attaches no semantics to *how* the files
+        in a File-set are to be structured so it's impossible to determine what
+        the layout will be when changes are to be made.
+
+        Parameters
+        ----------
+        ds_or_path : pydicom.dataset.Dataset, str or PathLike
+            An existing File-set's DICOMDIR, either as a pydicom
+            :class:`~pydicom.dataset.Dataset` or the path to the DICOMDIR file
+            as :class:`str` or :class:`pathlib.Path`.
+        include_orphans : bool, optional
+            If ``True`` (default) include instances referenced by orphaned
+            directory records in the File-set.
+        raise_orphans : bool, optional
+            If ``True`` then raise an exception if orphaned directory records
+            are found in the File-set (default ``False``).
+        """
+        if isinstance(ds_or_path, Dataset):
+            ds = ds_or_path
+        else:
+            ds = dcmread(ds_or_path)
+
+        sop_class = ds.file_meta.get("MediaStorageSOPClassUID", None)
+        if sop_class != MediaStorageDirectoryStorage:
+            raise ValueError(
+                "Unable to load the File-set as the supplied dataset is "
+                "not a 'Media Storage Directory' instance"
+            )
+
+        tsyntax = ds.file_meta.TransferSyntaxUID
+        if tsyntax != ExplicitVRLittleEndian:
+            warnings.warn(
+                "The DICOMDIR dataset uses an invalid transfer syntax "
+                f"'{tsyntax.name}' and will be updated to use 'Explicit VR "
+                "Little Endian'"
+            )
+
+        try:
+            path = Path(ds.filename).resolve(strict=True)
+        except FileNotFoundError:
+            raise FileNotFoundError(
+                "Unable to load the File-set as the 'filename' attribute "
+                "for the DICOMDIR dataset is not a valid path: "
+                f"{ds.filename}"
+            )
+        except TypeError:
+            # Custom message if DICOMDIR from bytes, etc
+            raise TypeError(
+                "Unable to load the File-set as the DICOMDIR dataset must "
+                "have a 'filename' attribute set to the path of the "
+                "DICOMDIR file"
+            )
+
+        self.clear()
+        self._id = ds.FileSetID
+        self._uid = ds.file_meta.MediaStorageSOPInstanceUID
```

Comment:
Should we raise if this is missing or workaround it? It's not critical, after all...